152/add set home command to set your own location

shen-sat opened the pull request:
Hey @JacobEvelyn - apologies for the delay in getting this done! Busy with the hustle and grid of life :)

I've added the implementation discussed in: https://github.com/JacobEvelyn/friends/issues/152

Summary of changes:

- `Event` now has an `implicit_location` attribute
- When `friends.md` is read, if an activity includes the words `moved to _LOCATION_`, then LOCATION is saved and applied as the `implicit_location` for any subsequent activities that don't have a location in their description (ie that don't have an explicit location).
- If another `moved to _LOCATION_ ` activity exists further along in the `friends.md` activity timeline, then this new LOCATION is set as the `implicit_location` for subsequent activities without an explicit location.
- If the activity `moved to _LOCATION_` does not appear in `friends.md`, then activities that don’t have an explicit location do not have an `implicit_location` set.

This feature has been tested to work with `friends list activities —in` and `friends list favourite locations` commands. 

We originally agreed on `to _LOCATION_` as the key words to trigger saving and applying `implicit_location`, but this would mean phrases like `went to _Marie's Diner_` would mistakenly get picked up and set as the `implicit_location` - hence I've changed the key words to `moved to _LOCATION_`, which is more explicit/specific.

I've redefined `CONTENT` specific for the tests - I tried to use the existing definition of `CONTENT`, but it was proving too tricky. I needed `CONTENT` that would have multiple `moved to _LOCATION_` lines to test that `implicit_location` is set in the correct way.

---

Hi there! Thanks so much for submitting a pull request!

Let's just make sure together that all of these boxes are checked before we
merge this change:

- [ ] The code in these changes works correctly.
- [ ] Code has tests as appropriate.
- [ ] Code has been reviewed by @JacobEvelyn.
- [ ] All tests pass on Travis CI.
- [ ] Code coverage remains high.
- [ ] Rubocop reports no issues on Travis CI.
- [ ] The `README.md` file is updated as appropriate.

Don't worry—this list will get checked off in no time!

JacobEvelyn commented on lib/friends/event.rb:
```diff
@@ -132,9 +133,21 @@ def update_location_name(old_name:, new_name:)
+      location_in_description?(location) || location_is_implicit?(location)
+    end
+
+    def location_in_description?(location)
```
Can this method be private?

shen-sat replied:
Yup! I think `location_is_implicit?` can be made private too

shen-sat replied:
Done

JacobEvelyn commented on lib/friends/event.rb:
```diff
@@ -132,9 +133,21 @@ def update_location_name(old_name:, new_name:)
+      (@description.scan(/(?<=_)[^_]+(?=_)/).include? location.name)
+    end
+
+    def location_is_implicit?(location)
```
Can this method be private?

shen-sat replied:
Done

JacobEvelyn commented on lib/friends/introvert.rb:
```diff
@@ -658,6 +658,19 @@ def set_n_activities!(type)
       end
     end
 
+    def set_implicit_locations(_activities)
```
Looks like this param can be removed?

shen-sat replied:
Yup! Done

JacobEvelyn commented on lib/friends/introvert.rb:
```diff
@@ -658,6 +658,19 @@ def set_n_activities!(type)
 
+    def set_implicit_locations(_activities)
+      implicit_location = nil
+      @activities.reverse_each do |activity|
```
Ah, using `reverse_each` is clever!

shen-sat replied:
Haha, this was a typo that happens to still work 😂

JacobEvelyn commented on test/commands/list/activities_spec.rb:
```diff
@@ -52,8 +52,39 @@
       end
-    end
 
+      describe "when implicit location is set" do
```
This test is very thorough, nice! Would you mind also adding our implicit location case to [this test](https://github.com/JacobEvelyn/friends/blob/master/test/commands/clean_spec.rb#L72) to ensure that if we've never seen the location before it gets added?

shen-sat replied:
Hey @JacobEvelyn, could you expand on what you mean by moving our implcit location here? Implicit locations don't seem to have anything to do with cleaning the `friends.md` file - any `moved to _LOCATION_` lines and lines with no location in their description would be read and treated exactly the same as the lines already included in this test, and any implcit locations set on any of the activities would not have an effect on what is stored in the `friends.md` file.

JacobEvelyn replied:
Ah, good question. So any time info is added we [call `clean` internally to re-write the file](https://github.com/JacobEvelyn/friends/blob/master/bin/friends#L97). And when a new location is mentioned in an added activity, it's [automatically added to the location list](https://github.com/JacobEvelyn/friends/issues/182). So I'd just like a test to confirm that this behavior still happens when the new location is first mentioned in a `moved to _New Place_` activity.

Does that make sense?

shen-sat replied:
I will push back slightly: from my understanding, a `moved to _LOCATION_` line will be treated by the `clean` method in exactly the same way as any other line containing _LOCATION_ (the regex cares about the _LOCATION_ part, and not the `moved to` part), hence perhaps there is no need to add a 'moved to _LOCATION_ line'. But, it is a small change and the test remains read-able so I've added this to the test :)

shen-sat replied:
Yup makes sense! I haven't thought about regression much in the projects I've worked on, but I guess there is balancing act - you don't want to write test absolutely everything in your when you add a feature, but you want a safety net when it is probable a feature could touch other parts of the code. I need more exposure to walking that fine line :)

JacobEvelyn commented on lib/friends/event.rb:
```diff
@@ -132,9 +133,21 @@ def update_location_name(old_name:, new_name:)
+    end
+
+    def location_in_description?(location)
+      (@description.scan(/(?<=_)[^_]+(?=_)/).include? location.name)
```
Since this is so close to the behavior of `location_names`, how would you feel about just using that here? e.g. `location_names.include? location.name`.

JacobEvelyn replied:
Also if we do that I think it makes sense to memoize the result of `location_names`, e.g.

```
def location_names
  @location_names ||= @description.scan(/(?<=_)\w[^_]*(?=_)/).uniq
end
```

since this is always called in `set_implicit_locations` and then also called again when filtering (e.g. `friends list activities --in Paris`) _or_ writing the file (the `clean` method in Introvert). But please check my thinking on that!

shen-sat replied:
> Since this is so close to the behavior of `location_names`, how would you feel about just using that here? e.g. `location_names.include? location.name`.

Done!

> Also if we do that I think it makes sense to memoize the result of `location_names`, e.g.
> 
> ```
> def location_names
>   @location_names ||= @description.scan(/(?<=_)\w[^_]*(?=_)/).uniq
> end
> ```
> 
> since this is always called in `set_implicit_locations` and then also called again when filtering (e.g. `friends list activities --in Paris`) _or_ writing the file (the `clean` method in Introvert). But please check my thinking on that!

Some tests are failing when I memoize the result of `location_names`, namely the test to do with renaming locations.
Whenever the `rename location` command is called, `the set_implicit_location` command is run first and the `clean` command runs at the end. Both of these methods call `location_names`.
When we memoize the result of `location_names` to `@location_names`, and run the `rename location` command, `set_implicit_location` runs `location_name` and sets `@location` to whatever the current location (eg London) in the description is. Then the current location in the description and `### Locations` section is renamed (to eg Paris). Then the `clean` command runs `location_names` command again to check for any disparities between locations in activities and locations in `### Locations` - and because `@locations` is set to London, the clean command will think there is a disparity and add London as a new location.

I'm going to therefore leave this as is, unless you have another suggestion?

JacobEvelyn replied:
Ah, interesting! Thanks for digging into that. I think you're right that leaving it as-is is a good way to proceed. (And wow am I glad this codebase has so many good tests!)

shen-sat replied:
Awesome - I'm glad there are loads of tests too, makes my task so much easier 😄

JacobEvelyn commented on lib/friends/introvert.rb:
```diff
@@ -658,6 +658,19 @@ def set_n_activities!(type)
+          next
+        end
+        # Does activity have a location in description?
+        activity.implicit_location = implicit_location if activity.location_names.empty?
```
I find it a little confusing to have `location_names` not include the implicit location. Could we rename it to something like `explicit_location_names` or `description_location_names` to make this clearer?

shen-sat replied:
Agree, I've changed to `description_location_names`

shen-sat replied:
I agree with you, this could be changed to something more explicit. I therefore tried changing all references to `location_names` to `description_location_names` - however, I got errors when running the tests, as something in the code was still somehow calling `location_names`. A little investigation with `pry` revealed this line:

**friends/lib/friends/introvert.rb:641)**
```
activity.send("#{type}_names").each do |thing_name|
``` 
We could split this one line into an `if` statement that explicity calls `description_location_names` if `type` is a `location` (else it will call `friends_names`). But I wanted to discuss with you whether that's the right thing to do first.

JacobEvelyn replied:
Ah, nice catch. Looking at that line, it's used in `set_n_activities!` which sets the `n_activities` field so that commands like `friends list favorite locations` works correctly. Since that should include the implicit locations, I think what makes the most sense is to rename the method to `description_location_names` as you had, and make a new `location_names` method that includes both explicit and implicit location names. Does that make sense?

This is also making me realize that we should add one or two new tests to `friends list favorite locations` to confirm this behavior (and also confirm that a location can't double-count a location if it's mentioned twice either explicitly or implicitly). Does that make sense and seem right?

shen-sat replied:
Done and done - `description_location_names` has been replaced the old `location_names`, and I've created a new `location_names` with a different implementation to take into account both locations in descriptions and implicit locations.
I've also added a test for `list favorite locations`. I've also removed the location `London` from the tests I've created, because I realised they it isn't necessary

JacobEvelyn commented on lib/friends/event.rb:
```diff
@@ -64,6 +64,7 @@ def initialize(str: "")
 
     attr_reader :date
     attr_accessor :description
+    attr_accessor :implicit_location
```
Looks like this could just be `attr_writer`, right?

shen-sat replied:
Ha, I've not always been super clear on the difference, but I just played around with `irb` and I understand now! Totally, it's only called once outside of `Event` to be assigned, so I can change this 👍

JacobEvelyn commented on friends.gemspec:
```diff
@@ -35,5 +35,6 @@ Gem::Specification.new do |spec|
 
   spec.add_development_dependency "minitest", "~> 5.5"
   spec.add_development_dependency "minitest-proveit", "~> 1.0"
+  spec.add_development_dependency "pry"
```
Totally fine to add whatever you like for development, of course (I add plenty of my own too!), but please remove this to keep the dependency list as clean as possible 😄

shen-sat replied:
Whoops, my bad! I will remove this 👍

JacobEvelyn commented on lib/friends/introvert.rb:
```diff
@@ -658,6 +658,19 @@ def set_n_activities!(type)
     end
 
+    def set_implicit_locations(_activities)
+      implicit_location = nil
```
I _think_ this line is unnecessary, but I could be getting my scoping rules mixed up.

shen-sat replied:
No you are correct, I'll remove this 👍

shen-sat replied:
No you are correct, I will remove

shen-sat replied:
It's needed at this level to be available at different scopes (removing it causes tests to fail), so I'll keep it in.

JacobEvelyn commented:
Just wanted to apologize for the delay; I'm traveling a bunch right now but will try to review this over the next few days if I can!

shen-sat commented:
Hey @JacobEvelyn, hope you are having fun on your travels! I've had lots of fun working on this PR via using `pry` 🤓

I've addressed most of your comments by making the relevant changes and pushing to this branch. However there are some where I need some more clarification, or that required a little discussion.

Thanks for giving me the heads-up that you're busy - I'll await your comments before proceeding :)

JacobEvelyn commented on lib/friends/introvert.rb:
```diff
@@ -677,6 +688,7 @@ def read_file
         # Parse the line and update the parsing state.
         state = parse_line!(line, line_num: line_num, state: state)
       end
+      set_implicit_locations
```
Ah, this is a minor thing but I've been using bangs (`!`) as a signal that objects are being mutated. Since this is setting a field on Locations, would you mind making this `set_implicit_locations!`?

shen-sat replied:
I read about bangs and `?` just the other day :) Makes sense and done!

shen-sat replied:
Weird - I made the change and it passes on my local, but not in Travis. I looked at the failing test and I can't quite figure out why it would fail after adding a bang to `set_implicit_locations` 🤔 Do you think you could have a look? I've reverted the commit for this for now while we figure out what is going on

JacobEvelyn commented:
Just left some more comments! Thanks for such thoughtful discussion!

shen-sat commented:
Hey @JacobEvelyn! Hope you're enjoying your travels :)

I've made all the changes you suggested (see my comments). I don't know why but some of your comments have been stated twice on the PR timeline, but you should see a reply/resolution from me to all of them (I've decided I really don't like Github's UI for PRs 😕. Have you used any other version control platform that has a better way of managing comments? I'm not suggesting it for this project, just curious 🙂).

Importantly, one commit is breaking the build and I don't know why (see comments). I've looked at the failing test, and it doesn't have anything to do with its associated commit. Would you mind taking a look?

JacobEvelyn commented:
Hey @shen-sat! Sorry for the delay!

> I don't know why but some of your comments have been stated twice on the PR timeline, but you should see a reply/resolution from me to all of them (I've decided I really don't like Github's UI for PRs 😕)

Ick, agreed! 😞 

> Have you used any other version control platform that has a better way of managing comments?

I used to use Bitbucket a lot for work, and while it certainly has its own quirks (the ecosystem is much worse, and in general it seems to be designed primarily for hosting Java/JVM code), the PR process is much nicer in many small ways. Unfortunately, I don't get to use it anymore 😭 

I've also heard good things about GitLab but never tried it.

> Importantly, the last commit is breaking the build and I don't know why (see comments). I've looked at the failing test, and it doesn't have anything to do with its associated commit. Would you mind taking a look? I've reverted the commit for now.

Really appreciate you digging in here and proactively reverting the commit as well! You are by far the most thoughtful and professional contributor I've had here! 😄

I don't see why that change would affect that test either, which is making me think it's a nondeterministic test failure. Of course I don't want nondeterministic test failures in general, but since it seems unrelated to your changes and I haven't seen it before, I think it makes sense for me to dig in separately but not block your changes on that test. In the meantime I re-ran the job on Travis on that commit and it passed (note now that that commit in GitHub now has the green check), so if you'd be so kind as to re-revert and get back to the `set_implicit_locations!` version I think we'll be good to merge this!

JacobEvelyn commented:
Oh also, as a heads up—to keep the git history clean I was planning to "squash and merge" or "rebase and merge" your PR and write a cleaner commit message. But if you'd rather do that on your end and squash all of the commits into one so I can just merge that works too! Whichever is easier for you is fine by me.

shen-sat commented:
Hey @JacobEvelyn, it's been too awesome working on this PR. I've learnt so much about Ruby and open-source etiquette 😄 And a lot of that has been down to you being a super helpful and considerate maintainer. I've really appreciated the space to have discussions and the chance to push back on things, even though I'm new to your project and new to OS in general.

I will let you figure out the non-deterministic test issue, and I'm happy to squash my commits into one with a useful commit message. And once you merge this will be my first feature contribution to open-source -  woop! 🙌 

Two more things:
- Do you have any constructive feedback for me? I'd like to continue to contribute to OS, so any feedback to help me improve my code/interaction with maintainers would be greatly appreciated 👍 
- I'd quite like to pick up another issue for this project if that's cool with you (maybe after a little break)? I know the code base fairly well-ish now, so it'd be a shame if I didn't put that knowledge to use again. Plus I love the app and all its tests! 🤓

shen-sat commented on lib/friends/introvert.rb:
```diff
@@ -658,6 +658,17 @@ def set_n_activities!(type)
+      @activities.reverse_each do |activity|
+        if activity.moved_to_location
+          implicit_location = activity.moved_to_location
+          next
```
Done! I'm not sure why that was there originally 🤔
